File: spark-batch/src/main/kotlin/com/sumo/experiments/LoadJob.kt
package com.sumo.experiments

import org.apache.spark.sql.SaveMode.*
import org.apache.spark.sql.functions.*
import org.apache.spark.sql.SparkSession

fun main(args: Array<String>) {


    val spark = with(SparkSession.builder()) {
        appName("Java Spark SQL basic example")
        config("mapreduce.fileoutputcommitter.marksuccessfuljobs", "false")
        config("parquet.enable.summary-metadata", "false")
        getOrCreate()
    }

    spark.sparkContext().setLogLevel("ERROR")

    val input = spark.sparkContext().conf["spark.myapp.input"]
    val output = spark.sparkContext().conf["spark.myapp.output"]

    val people = with(spark.read()) {
        option("header", "true")
        // option("inferSchema", "true")
        schema(Schema.peopleSchema)
        option("nullValue", "null")
        option("mode", "DROPMALFORMED")
        csv("${input}/people.csv")
    }.cast<Person>()

    people.createOrReplaceTempView("people")

    val skills = with(spark.read()) {
        option("header", "true")
        // option("inferSchema", "true")
        schema(Schema.skillSchema)
        option("nullValue", "null")
        option("mode", "DROPMALFORMED")
        csv("${input}/skills.csv")
    }.cast<Skill>()
    skills.createOrReplaceTempView("skills")

//    val result = spark.sql("SELECT age, count(*) as cnt  FROM people WHERE age >= 13 AND age <= 19 group by age")
//    val result = spark.sql("SELECT id, name, numFriends as friends, age  FROM people WHERE age >= 13 AND age <= 19")
    val result = spark.sql("""SELECT people.*, collect_list(struct(skills.*)) AS skills
                                                    FROM people LEFT JOIN skills ON  people.id =  skills.pid
                                                    GROUP BY id, name, age, numFriends
                                                    ORDER BY id""")

    println("--------------")
    println(result.printSchema())
    result.show(5, false)

    result.coalesce(1)
        .write()
        .mode(Overwrite)
        .json("${output}/people.json")
    println("--------------")
}

File: spark-batch/src/main/scala/com/sumo/experiments/scala/LoadJob.scala
package com.sumo.experiments.scala

import org.apache.spark.sql.SparkSession

object LoadJob extends App {
    //  println("Hello, " + args(0))

    val spark = SparkSession.builder()
          .appName("People-Skills Join Job")
          .config("mapreduce.fileoutputcommitter.marksuccessfuljobs", "false")
          .config("parquet.enable.summary-metadata", "false")
          .getOrCreate()

    import spark.implicits._

    val sc = spark.sparkContext

    sc.setLogLevel("ERROR")

    val input = sc.getConf.get("spark.myapp.input")
    val output = sc.getConf.get("spark.myapp.output")

    val people = spark.read
        .option("header","true")
        // .option("inferSchema", "true")
        .schema(Schema.peopleSchema)
        .option("nullValue", "null")
        .option("mode", "DROPMALFORMED")
        .csv(s"$input/people.csv").as[Person]

    people.createOrReplaceTempView("people")

    val skills = spark.read
        .option("header","true")
        // .option("inferSchema", "true")
        .schema(Schema.skillSchema)
        .option("nullValue", "null")
        .option("mode", "DROPMALFORMED")
        .csv(s"$input/skills.csv").as[Skill]

     skills.createOrReplaceTempView("skills")

    val result = people.as("L").joinWith(skills.as("R"), $"L.id" === $"R.pid",  "left_outer" )

    val scalaGuys = result
        .filter( p => p._1.age > 26 && p._2!=null && p._2.skill.contentEquals("scala") )
        .map(p => (p._1.name, p._1.age, p._1.numFriends))


    println("--------------")
    result.explain
    println(s"partitions: ${result.rdd.partitions.length}")
    println(result.printSchema)
    result.show(20, truncate = false)
    println(result.count)
    scalaGuys.show(20, truncate = false)
    println("--------------")
}
